Prevent reordering node above its check node

ehsankianifar opened the pull request:
Reordering should not move a node above where its children are checked.
@hzongaro

ehsankianifar commented on compiler/optimizer/LocalReordering.cpp:
```diff
@@ -694,9 +694,7 @@ bool TR_LocalReordering::isAnySymInDefinedBy(TR::Node *node, vcount_t visitCount
 
-      if ((!node->getOpCode().isLoadVar() ||
-           node->mightHaveVolatileSymbolReference()) &&
-          !node->getOpCode().isCheck())
```
Removing the condition to bypass check nodes.

ehsankianifar commented on compiler/optimizer/LocalReordering.cpp:
```diff
@@ -407,11 +407,9 @@ bool TR_LocalReordering::insertEarlierIfPossible(TR::TreeTop *storeTree, TR::Tre
-      stopHere = true;
-   }
+      // Moving a node earlier than a check may have unforeseeable consequences as sophisticated optimizations like global vp can perform transformations based on value numbers.
+      if (!stopHere && currentNode->getOpCode().isCheck())
```
@hzongaro @vijaysun-omr it was blocking move above checkCast before. I changed it to block moving for all check nodes. Please let me know if it looks good to you. Thanks.

ehsankianifar commented:
All sanity.functional tests passed on arm 64 linux, z linux, and x86-64 linux with java 11 and 21.
java 8 builds and functional tests passed on zos, z linux, and x86 linux.
manually tested and confirm that the wrong reordering is not happening with this fix.

vijaysun-omr commented on compiler/optimizer/LocalReordering.cpp:
```diff
@@ -430,7 +428,7 @@ bool TR_LocalReordering::insertEarlierIfPossible(TR::TreeTop *storeTree, TR::Tre
             storeTree->setNextTreeTop(nextTree);
             nextTree->setPrevTreeTop(storeTree);
-      }
+            }
```
Indentation is not right (before the change, but also after it).

ehsankianifar replied:
This closing bracket does not belong to `if (stopHere)`. there is a nested if inside it which this bracket belongs to.

vijaysun-omr commented:
Small typo in commit message : "case" should be "cause"

vijaysun-omr commented:
Small typo in commit message : "abode" should be "above"

ehsankianifar commented:
thanks @vijaysun-omr for your review. I changed the commit message. Also added link to the github issue that it fix but not sure if it was okay since the issue is on openj9 repo. Please let me know if it looks good to you.

vijaysun-omr commented:
I approved, but will wait for @hzongaro to review before I launch more testing.

hzongaro reviewed (approved):
I think the change looks good.  Thanks for tracking down the problem!

vijaysun-omr commented:
Jenkins build all

vijaysun-omr commented:
Checks have passed except zos. Reviews are also done. Merging.
